python/lsst/daf/butler/dimensions/record_cache.py

```diff
@@ -0,0 +1,157 @@
+
+    @contextmanager
+    def modifying(self, element: str) -> Iterator[DimensionRecordSet | None]:
+        """Return a context manager for modifying the cache and database
+        content consistently.
+
+        Parameters
+        ----------
+        element : `str`
+            Name of the dimension element whose records will be modified.
+            If this is not a cached record, `None` will be returned and the
+            context manager does nothing.
+
+        Returns
+        -------
+        context : `contextlib.AbstractContextManager` [ `DimensionRecordSet` \
+                or `None` ]
+            A context manager that when entered returns a `DimensionRecordSet`
+            that should be modified in-place, or `None` if the cache is
+            currently reset.
+
+        Notes
+        -----
+        The returned context manager resets the cache when entered, and only
+        restores the cache (along with the modifications to the returned
+        `DimensionRecordSet`) if an exception is not raised during the context.
+        It also takes care of updating any cache for "implied union" dimensions
+        (e.g. ``band``, in the default dimension universe) when their targets
+        are updated (e.g. ``physical_filter``).
+        """
+        if element in self:
+            records = self._records
+            self._records = None
+            if records is None:
+                yield None
+            else:
+                yield records[element]
+                for other_element_records in records.values():
+                    other_element = other_element_records.element
+                    # If we've just updated the records of a dimension element
+                    # that is the implied union target of another (i.e. we've
+                    # updated physical_filter, and thus possibly updated the
+                    # set of band values).  We need to update the cache for
+                    # the implied union target (i.e. band), too.
+                    if (
+                        other_element.implied_union_target is not None
+                        and other_element.implied_union_target.name == element
+                    ):
+                        other_element_records.update(
+                            other_element.RecordClass(
+                                **{other_element.name: getattr(record, other_element.name)}
+                            )
+                            for record in records[element]
+                        )
+            self._records = records
+        else:
+            yield None
+
+    def load_from(self, other: DimensionRecordCache) -> None:
+        """Load records from another cache, but do nothing if it doesn't
+        currently have any records.
+
+        Parameters
+        ----------
+        other : `DimensionRecordCache`
+            Other cache to potentially copy records from.
+        """
+        self._records = copy.deepcopy(other._records)
```

Comment:
This cache is a bit problematic for the service backend use case in its current form, but I'm happy to bang on it myself in the ticket I'm working on.  This refactor is a huge improvement over the existing code already.

The most immediate (and easy) issue is that services instantiate a "template" Butler object that never does any work, so this copy never has anything to copy.  Thus we end up re-fetching the cache on every request.  Not necessarily worth worrying about, but I would also prefer that this not do so much copying -- was hoping that cloning a Butler would be on the order of "instantiating a few empty dicts and small objects and copying a few pointers".  

Two possible ways to fix that:
1. Force a 'prefetch' in LabeledButlerFactory so the data is there to copy.
2. Make this cache threadsafe and shareable between instances.

I might go ahead and remove `__getitem__` and add `find` as I described in the comment above.  This would allow us to recover from someone inserting dimension records in the background, without a server restart. With `find` called from this class instead of directly on the `DimensionRecordSet`, we can detect the cache miss and force a refresh.

There are some problems with the interaction between multiple Butler instances and this cache.  This is "mostly theoretical" because these dimensions are "mostly immutable", but I think worth mentioning because the same issues come up with any Butler-related cache.  These issues can be ignored at the moment for read-only Butler server, but I'll need to make sure whatever I do here doesn't cause breakage for usage of DirectButler cloned through the same function.

If the cache is NOT shared, any mutation of the dimension records causes the cache in any other Butler instance to become stale.  (Of course this also applies to a mutation done by a separate process entirely.)

If the cache IS shared, the big issue is that we allow long-lived user-initiated "outer" transactions, so it's tough to know when it's safe to update the cache.  If we have mutated the data in one Butler instance, it may be rolled back before becoming "real", and we end up out of sync with the server.

Even doing the initial fetch or a refresh on a cache miss is potentially problematic if the cache is shared.  Depending how old the transaction is, we may "move backwards in time" from another instance's point of view if we force-refresh the entire cache.  Even if there are no transactions, with multiple threads a delayed response from the database could have the same effect.

So I think my approach might be:
* Keep the cache copied between instances instead of shared.
* For the server, prefetch everything when the "template" instance is created.
* Move `find` onto this class
* Read through to `fetch_one` on a cache miss instead of throwing an error, and update our non-shared copy of the cache.

We're still hosed if someone updates a cached dimension record, but at least inserts work with only a small performance penalty.

Reply:
> You're welcome to add `find` to this class on your ticket branch instead of me doing it on DM-42344.

I think I'll probably hold off on this for now, since this cache is truly immutable for the "fully read-only data release DB" use case I'm trying to get working right now.

>     * I eventually want to get rid of the possibility of user-defined outer transactions in favor of having a way for users to batch up a number of operations to be performed atomically in a more declarative sense.

Yeah I like that approach, should be a huge win if we can make it work.
 
>     * `expandDataId` calls are all over the place

Thanks for calling this out -- I was vaguely aware of the existence of the expanded data ID thing but hadn't done any thinking on the implications yet.  I'll leave the change to reading through on a miss for later as well then, since I don't need it for my immediate use case.

>     * We really want this cache to be client-side in `RemoteButler` for the query-result-objects use case (where misses are not a concern);

I think we can allow this.  For cases like this where it's like "~50MB or smaller, almost never changes" we can use a pattern like:
1.  Client downloads the full cache on first use.
2. Server sends a hash or version number for the cache with every query response, so the client knows when it  needs to refresh the cache.  (In this case I think we would pre-generate a sorted, JSON-serialized version of the cache data at startup and generate an SHA1 of the JSON string.)
3. Handle cache invalidation on the server "somehow".  (For this particular case we can probably just manually restart the server, since these changes are so infrequent.)

Because the cache is versioned we can safely share it between threads or even write it to local disk for use between runs.  This same pattern would work OK for the DimensionUniverse cache as well.